package-res/resources/web/pentaho/visual/base/model.js

```diff
@@ -45,37 +73,300 @@ define([
+          return this._broadcastResult(DidSelect, RejectedSelect, result, dataFilter);
+        },
+
+        /**
+         * Updates the current selection filter based on a filter and on a selection mode.
+         *
+         * This method is invoked by [selectAction]{@link pentaho.visual.base.Model#selectAction} once all
+         * listeners of the ["will:select"]{@link pentaho.visual.base.events.WillSelect} event have eventually
+         * replaced the original input filter and the selection mode to be used.
+         *
+         * @param {!pentaho.data.filter.AbstractFilter} inputDataFilter - A filter representing
+         * the data set which will be used to modify the current selection.
+         * @param {!function} selectionMode - A function that represents how the selection made by the user
+         * will be merged with the current selection.
+         *
+         * @return {pentaho.lang.ActionResult}
+         * If unsuccessful, the `error` property describes what originated the error.
+         * If successful,  the `error` property is `null` and the `value` property contains the updated current selection filter.
+         *
+         * @fires "will:change:selectionFilter"
+         * @fires "did:change:selectionFilter"
+         * @fires "rejected:change:selectionFilter"
+         *
+         * @see #selectAction
+         * @see #_runPipeline
+         * @see #_broadcastResult
+         * @protected
+         */
+        _changeSelection: function(inputDataFilter, selectionMode) {
+          var dataFilter = inputDataFilter;
+          var result = this._runPipeline(WillChangeSelection, dataFilter,  null);
+
+          var noError = result && !result.error;
+          if(noError) {
+            dataFilter = result.value.dataFilter;
+          }
+          if(noError || !result) {
+            try {
+              var currentFilter = this.get("selectionFilter");
+              var combinedSelection = selectionMode(currentFilter, dataFilter);
+              if(combinedSelection) {
+                this.set("selectionFilter", combinedSelection);
+                result = ActionResult.fulfill(combinedSelection);
+              } else {
+                result = ActionResult.reject(new UserError(bundle.structured.error.selection.invalid));
               }
-            ]
+            } catch(e) {
+              result = ActionResult.reject(e);
+            }
+          }
+
+          return this._broadcastResult(DidChangeSelection, RejectedChangeSelection, result, dataFilter);
+        },
+
+        /**
+         * Executes an action when the user interacts with a visual element, normally by double clicking it.
+         *
+         * The flow starts by triggering the event
+         * {@link pentaho.visual.base.events.WillExecute|will:execute}.
+         * Its event listeners can be attributed a _priority_,
+         * and can be regarded as operations in a processing pipeline that are allowed to:
+         * - cancel the event,
```

Comment:
- cancel the event
- replace the input data filter
- change the `doExecute` action